### ISSUE TITLE
El corrector deja de procesar entregas si un TP crea archivos sin parar.

### ISSUE THREAD

**Issue by ndvazquez** [algoritmos-rw/corrector#17]
Aparentemente si un alumno envía una entrega en la que por error se crean archivos de 0 bytes dentro de un `while (true)`, el worker que lo ejecuta se cuelga y no se "mata" (perdón por los términos burdos). A causa de esto fetchmail deja de levantar mails y se nos "cuelga" el corrector.

@sportelliluciano sugirió que el problema es que se debe estar quedando sin espacio en la tabla de inodes y parece ser esto.

Antes de ejecutar el TP:
```
$ df --inodes
Filesystem       Inodes  IUsed   IFree IUse% Mounted on
udev             974766    595  974171    1% /dev
tmpfs            982171    941  981230    1% /run
/dev/sda5      10059776 841725 9218051    9% /
```
Y 10 segundos después...
```
$ df --inodes
Filesystem       Inodes   IUsed   IFree IUse% Mounted on
udev             974766     595  974171    1% /dev
tmpfs            982171     941  981230    1% /run
/dev/sda5      10059776 1638916 8420860   17% /
```

TP que trajo a la luz esto -> https://github.com/algoritmos-rw/algo2_entregas/blob/a4ecac2d21ee97826622bcdb20f95ad6a0b96d2b/tp2/2018_1/100541_100948/analog.c

**Comment by dato** [comment]
No entiendo nada de este report.

El corrector corre el código de los alumnos [con un `tmpfs` montado en `/tmp`](https://github.com/algoritmos-rw/corrector/commit/a3bdb6aa2acb34548700609a8470989db5da5142); el resto del filesystem se monta como sólo lectura. Así que no entiendo cómo se va a estar comiendo los ínodos del sistema. Además, ¿en qué host corrió ese `df`? No hay `/dev/sda5` en turing.

Acabo de enviar un TP2 con el siguiente _algueiza.c_:

```c
#include <stdlib.h>

int main(void) {
    system("while true; do mktemp; done");
}
```

La respuesta fue:

```
ERROR: El proceso tardó más de 120 segundos

rm -rf algueiza  *.o *.d *.dSYM
cc -g -std=c99 -Wall -Wtype-limits -pedantic -Wconversion -Wno-sign-conversion -Werror   -c -o algueiza.o algueiza.c
cc -g -std=c99 -Wall -Wtype-limits -pedantic -Wconversion -Wno-sign-conversion -Werror algueiza.o -o algueiza 
cd pruebas && ./pruebas.sh ../algueiza
Ejecución de pruebas unitarias:

01 Prueba: agregar_un archivo
```

A menos que alguien pueda decirme cómo reproducir el problema, creo que este bug es para cerrar. (Quizá convenga volver a enviar el ZIP de 100541 de nuevo, perdón pero no lo he hecho porque imagino que el _deps.mk_ no encaja, etc. Habría que rearmar el ZIP.)

**Comment by ndvazquez** [comment]
Recién llevé `skel` al commit con las pruebas del TP2 de 2018_1 e intenté realizar esta misma entrega, y efectivamente deja de responder el corrector. Como bien decís no se come inodes, desconocía lo de `tmpfs`, mala mía por no investigar apropiadamente.

Luego borré (sí, un asco perdón) el mail del mal en la casilla del corrector, reinicié el servicio e intenté dos entregas diferentes y respondió sin problemas.

Al enviar la entrega de 100541, el proceso del worker se queda colgado aunque no sabría bien por qué, pero asumo que está relacionado con lo de crear archivos de tamaño 0 ya que una vez que 100541 solucionó el bug que creaba archivos sin parar, el corrector le tomó bien la entrega y no hubo más dramas.

Lo que veo con systemctl al realizar la entrega de 100541 es lo siguiente

```
nvazquez@turing:/srv/apps/corrector-algo2/data/skel$ sudo systemctl status corrector@algo2.service
● corrector@algo2.service - Corrector automático para algo2
   Loaded: loaded (/etc/systemd/system/corrector@.service; enabled; vendor preset: enabled)
   Active: active (running) since Tue 2019-01-08 19:08:15 -03; 5min ago
 Main PID: 17840 (fetchmail)
    Tasks: 18 (limit: 4915)
   CGroup: /system.slice/system-corrector.slice/corrector@algo2.service
           ├─17840 /usr/bin/fetchmail --daemon 180 --nodetach --mda /srv/apps/corrector-algo2/bin/corrector
           ├─17841 sh -c /srv/apps/corrector-algo2/bin/corrector
           ├─17842 python3.6 /srv/apps/corrector-algo2/bin/corrector
           └─17843 docker run --rm --interactive --net none --env LANG=C.UTF-8 --memory 512m --memory-swap 512m --user nobody:nogroup --cap-drop ALL --read-only --tmpfs /tmp:exec,size=75M --ru
```

Y estuvo así un buen rato hasta que borré el mail y reinicié el servicio. Parece que no se termina el proceso del worker incluso pasados los 120 segundos (sería el que empieza con `docker run -rm blablabla`, correcto?).

Si es algo muy falopa y particular de la entrega de 100541 entonces mejor cerrar esto, pero me preocupaba un poco que nos volviera a pasar algo similar durante la cursada.

**Comment by dato** [comment]
No, si hay un ZIP que cuelga al corrector, eso es un bug y definitivamente se debe solucionar. Perdón si soné brusco en mi respuesta sobre los ínodos.

**Comment by dato** [comment]
El retraso lo introduce la llamada a `shutil.rmtree()` para borrar el directorio temporal creado. Es **MUY** lento. (Si se la sustituye con `rm -rf`, se bajan dos órdenes de magnitud el tiempo de ejecución.)

El issue de fondo que hay que arreglar es #28: no se aborta al _worker_ tras un timeout porque el que cuelga es código Python nuestro, no el compilado de los alumnos. En la implementación actual solo se controla este último.

Dejo este issue abierto porque, ortogonalmente a arreglar #28, habría que determinar si realmente nos hace falta borrar los contenidos del directorio temporal creado: como expliqué arriba, todos los archivos se crean en un _tmpfs_ efímero, lo cual quiere decir que desaparecen al terminar el container. Cualquier tiempo empleado en _cleanup()_ es tiempo perdido.

No obstante, sí me gustaría que el _worker_ se comporte bien si se ejecuta afuera de Docker (por ejemplo, en una de nuestras máquinas), en cuyo caso sí debería borrar los archivos temporales.

Mi propuesta sería introducir un flag `--no-cleanup`.
